create_mb_refstate script is failing for measured boot log from Ubuntu machine (VM and also physical - HPE)

# Environment
 * OS / version: Ub22.04
 * Processor architecture: N/A
 * TPM Manufacturer: N/A
 * Keylime version: 6.5.2

# Description 
Running create_mb_refstate script with measured boot log from Ubuntu 22.04 machine (both virtual and physical - HPE) fails with same reason (see in relevant logs)

## Expected behavior vs. actual behavior 
Expected:
measured_boot_reference_state.json file is properly created from input file (measured boot log)


## Steps to reproduce problem
1. run script as shown below in relevant logs section

# Relevant logs
Failure example:
server ~# ./create_mb_refstate /home/sysadmin/binary_bios_measurements measured_boot_reference_state.json
Reading configuration from ['/etc/keylime/logging.conf']
Reading configuration from ['/etc/keylime/tenant.conf']
2022-11-15 14:47:40.901 - keylime.tpm - INFO - TPM2-TOOLS Version: 5.2
Traceback (most recent call last):
 File "/root/./create_mb_refstate", line 173, in <module>
   main()
 File "/root/./create_mb_refstate", line 167, in main
   **get_kernel(events),
 File "/root/./create_mb_refstate", line 87, in get_kernel
   **assert len(out) in [3, 4], "Expected 3 different UEFI applications to be booted (Shim, Grub, Linux)"
AssertionError: Expected 3 different UEFI applications to be booted (Shim, Grub, Linux)**
server ~#

Attach any relevant log files that can help to debug your issue.

--- a/keylime/elchecking/example.py
+++ b/keylime/elchecking/example.py
@@ -67,6 +67,17 @@ allowed_kernel_test = tests.obj_test(
 
 allowed_kernel_list_test = tests.list_test(allowed_kernel_test)
 
+shim_authcode_sha256_no_secureboot = tests.obj_test(
+    shim_authcode_sha256=hex_test,
+    grub_authcode_sha256=hex_test,
+    vmlinuz_plain_sha256=hex_test,
+    initrd_plain_sha256=hex_test,
+    kernel_cmdline=tests.type_test(str),
+)
+
+
+allowed_kernel_list_test_no_secureboot = tests.list_test(shim_authcode_sha256_no_secureboot)
+
 
 class Example(policies.Policy):
     relevant_pcr_indices = frozenset(list(range(10)) + [14])
@@ -80,10 +91,15 @@ class Example(policies.Policy):
         if not isinstance(refstate, dict):
             raise Exception(f"Expected refstate to be a Python dict but instead got this Python value: {refstate!r}")
 
+        has_secureboot = refstate.get("has_secureboot", True)
+
         kernels = refstate.get("kernels")
         if not isinstance(kernels, list):
             raise Exception(f"refstate['kernels'] is {kernels!r} instead of a list")
-        allowed_kernel_list_test(kernels)
+        if has_secureboot:
+            allowed_kernel_list_test(kernels)
+        else:
+            allowed_kernel_list_test_no_secureboot(kernels)
 
         scrtm_and_bios_spec = refstate.get("scrtm_and_bios")
         if scrtm_and_bios_spec is None:
@@ -111,12 +127,20 @@ class Example(policies.Policy):
         vd_authority = tests.VariableDispatch()
 
         def bsa_test(kernel: typing.Dict[str, str]) -> tests.Test:
+            if not has_secureboot:
+                # When SecureBoot is disabled, GRUB does not use SHIM_LOCK to verify the signature therefore
+                # there is no EV_EFI_BOOT_SERVICES_APPLICATION entry.
+                tt = [
+                    tests.DigestTest({"sha256": string_strip0x(kernel["shim_authcode_sha256"])}),
+                    tests.DigestTest({"sha256": string_strip0x(kernel["grub_authcode_sha256"])}),
+                ]
+                return tests.TupleTest(*tt)
+
             tt = [
                 tests.DigestTest({"sha256": string_strip0x(kernel["shim_authcode_sha256"])}),
                 tests.DigestTest({"sha256": string_strip0x(kernel["grub_authcode_sha256"])}),
                 tests.DigestTest({"sha256": string_strip0x(kernel["kernel_authcode_sha256"])}),
             ]
-
             # In some scenarios the kernel gets measured twice
             tt2 = [
                 tests.DigestTest({"sha256": string_strip0x(kernel["shim_authcode_sha256"])}),
@@ -125,6 +149,13 @@ class Example(policies.Policy):
                 tests.DigestTest({"sha256": string_strip0x(kernel["kernel_authcode_sha256"])}),
             ]
             return tests.Or(tests.TupleTest(*tt), tests.TupleTest(*tt2))
+
+        def vmlinuz_test(kernel: typing.Dict[str, str]) -> tests.Test:
+            # Only when SecureBoot is disabled validate the vmlinuz digest measured by GRUB. When SecureBoot is enabled
+            # this was already done using the tests for EV_EFI_BOOT_SERVICES_APPLICATION.
+            if has_secureboot:
+                return tests.AcceptAll()
+            return tests.DigestTest({"sha256": string_strip0x(kernel["vmlinuz_plain_sha256"])})
 
         events_final = tests.DelayToFields(
             tests.And(
@@ -135,6 +166,7 @@ class Example(policies.Policy):
                             ipl9s=tests.TupleTest(
                                 tests.DigestTest({"sha256": string_strip0x(kernel["initrd_plain_sha256"])})
                             ),
+                            vmlinuz=tests.TupleTest(vmlinuz_test(kernel)),
                             kernel_cmdlines=tests.TupleTest(
                                 tests.RegExp("kernel_cmdline: " + kernel["kernel_cmdline"])
                             ),
@@ -147,6 +179,7 @@ class Example(policies.Policy):
             "kernel_cmdlines",
             "bsas",
             "ipl9s",
+            "vmlinuz",
             "s_crtms",
             "platform_firmware_blobs",
         )
@@ -155,9 +188,11 @@ class Example(policies.Policy):
         dispatcher.set((0, "EV_S_CRTM_VERSION"), events_final.get("s_crtms"))
         dispatcher.set((0, "EV_EFI_PLATFORM_FIRMWARE_BLOB"), events_final.get("platform_firmware_blobs"))
         dispatcher.set((7, "EV_EFI_VARIABLE_DRIVER_CONFIG"), vd_driver_config)
+        secure_boot_test = tests.FieldTest("Enabled", tests.StringEqual("Yes"))
+        if not has_secureboot:
+            secure_boot_test = tests.FieldTest("Enabled", tests.StringEqual("No"))
         # tpm2-tools versions < 5.2 parsed the GUIDs wrong therefore we include a check for both here
         # For more information see: https://github.com/keylime/keylime/issues/1003
-        secure_boot_test = tests.FieldTest("Enabled", tests.StringEqual("Yes"))
         vd_driver_config.set("61dfe48b-ca93-d211-aa0d-00e098032b8c", "SecureBoot", secure_boot_test)
         vd_driver_config.set("8be4df61-93ca-11d2-aa0d-00e098032b8c", "SecureBoot", secure_boot_test)
         pk_test = tests.OnceTest(
@@ -297,7 +332,10 @@ class Example(policies.Policy):
                     "Event", tests.FieldTest("String", tests.RegExp(r".*/loader/entries.*"))
                 ),  # Ignore  Boot Loader Spec files
                 tests.FieldTest("Event", tests.FieldTest("String", tests.RegExp(r".*/grub.*"))),
-                tests.FieldTest("Event", tests.FieldTest("String", tests.RegExp(r".*/vmlinuz.*"))),
+                tests.And(
+                    tests.FieldTest("Event", tests.FieldTest("String", tests.RegExp(r".*/vmlinuz.*"))),
+                    events_final.get("vmlinuz"),  # Note: this evaluates always to true if SecureBoot is enabled
+                ),
                 tests.And(
                     tests.FieldTest(
                         "Event",

--- a/scripts/create_mb_refstate
+++ b/scripts/create_mb_refstate
@@ -9,12 +9,15 @@ A simple script to generate a valid mb_refstate for the example policy
 
 import argparse
 import json
+import logging
 import re
 import sys
 
 from packaging.version import Version
 
 from keylime.tpm import tpm_main
+
+logging.basicConfig(level=logging.INFO, format="%(levelname)s - %(message)s")
 
 
 def event_to_sha256(event):
@@ -75,7 +78,7 @@ def get_keys(events):
     return out
 
 
-def get_kernel(events):
+def get_kernel(events, secure_boot):
     """
     Extract digest for Shim, Grub, Linux Kernel and initrd.
     """
@@ -84,13 +87,13 @@ def get_kernel(events):
         if event["EventType"] != "EV_EFI_BOOT_SERVICES_APPLICATION":
             continue
         out.append(event_to_sha256(event))
-    assert len(out) in [3, 4], "Expected 3 different UEFI applications to be booted (Shim, Grub, Linux)"
 
-    kernel = {
-        "shim_authcode_sha256": out[0]["sha256"],
-        "grub_authcode_sha256": out[1]["sha256"],
-        "kernel_authcode_sha256": out[2]["sha256"],
-    }
+    kernel = {"shim_authcode_sha256": out[0]["sha256"], "grub_authcode_sha256": out[1]["sha256"]}
+    if secure_boot:
+        assert len(out) in [3, 4], "Expected 3 different UEFI applications to be booted (Shim, Grub, Linux)"
+        kernel["kernel_authcode_sha256"] = out[2]["sha256"]
+    else:
+        assert len(out) == 2, "Expected 2 different UEFI applications to be booted (Shim, Grub)"
 
     for event in events:
         if event["EventType"] != "EV_IPL" or event["PCRIndex"] != 9:
@@ -98,6 +101,15 @@ def get_kernel(events):
         if "initrd" in event["Event"]["String"] or "initramfs" in event["Event"]["String"]:
             kernel["initrd_plain_sha256"] = event_to_sha256(event)["sha256"]
             break
+
+    if not secure_boot:
+        logging.info("Adding plain sha256 digest of vmlinuz for GRUB to refstate, because SecureBoot is disabled")
+        for event in events:
+            if event["EventType"] != "EV_IPL" or event["PCRIndex"] != 9:
+                continue
+            if "vmlinuz" in event["Event"]["String"]:
+                kernel["vmlinuz_plain_sha256"] = event_to_sha256(event)["sha256"]
+                break
 
     for event in events:
         if event["EventType"] != "EV_IPL" or event["PCRIndex"] != 8:
@@ -125,6 +137,14 @@ def get_mok(events):
     return out
 
 
+def secureboot_enabled(events):
+    for event in events:
+        if event["EventType"] == "EV_EFI_VARIABLE_DRIVER_CONFIG" and event["Event"].get("UnicodeName") == "SecureBoot":
+            return event["Event"]["VariableData"]["Enabled"] == "Yes"
+
+    raise Exception("SecureBoot state could not be determined")
+
+
 def main():
     tpm = tpm_main.tpm()
 
@@ -140,6 +160,12 @@ def main():
         help="Binary UEFI eventlog (Normally /sys/kernel/security/tpm0/binary_bios_measurements)",
     )
     parser.add_argument(
+        "--without-secureboot",
+        "-i",
+        action="store_true",
+        help="Set if you want to create a mb_refstate without SecureBoot (only MeasuredBoot)",
+    )
+    parser.add_argument(
         "mb_refstate", type=argparse.FileType("w"), help="Output path for the generated measure boot reference state."
     )
     args = parser.parse_args()
@@ -147,15 +173,29 @@ def main():
 
     failure, log_data = tpm.parse_binary_bootlog(log_bin)
     if failure or not log_data:
-        print(f"Parsing of binary boot measurements failed with: {list(map(lambda x: x.context, failure.events))}")
+        logging.error(
+            f"Parsing of binary boot measurements failed with: {list(map(lambda x: x.context, failure.events))}"
+        )
         sys.exit(1)
 
     events = log_data.get("events")
     if not events:
-        print("No events found on binary boot measurements log")
+        logging.error("No events found on binary boot measurements log")
         sys.exit(1)
 
+    has_secureboot = secureboot_enabled(events)
+    if not has_secureboot and not args.without_secureboot:
+        logging.error("Provided eventlog has SecureBoot disabled, but -i flag was not set")
+        sys.exit(1)
+
+    if has_secureboot and args.without_secureboot:
+        logging.warning(
+            "-i was set to create a refstate without SecureBoot, "
+            "but he provided eventlog has SecureBoot enabled. Ignoring this flag!"
+        )
+
     mb_refstate = {
+        "has_secureboot": has_secureboot,
         "scrtm_and_bios": [
             {
                 **get_s_crtm(events),
@@ -164,7 +204,7 @@ def main():
         ],
         **get_keys(events),
         **get_mok(events),
-        **get_kernel(events),
+        **get_kernel(events, has_secureboot),
     }
     json.dump(mb_refstate, args.mb_refstate)
 

